### ISSUE TITLE
respect open fd limit

### ISSUE THREAD

**Issue by jbenet** [ipfs/go-ipfs#1456]
bad things happen when we hit the fd limit. we should respect it. what we should do is:

- [ ] make flatfs respect ulimit by using a pkg level sema on open files? or something.
- [ ] improve https://github.com/jbenet/go-reuseport to not consume 1fd per conn (or fix it in go...)
- [ ] make sure we dont leak / leave unused fds open
- [ ] if limit < 1025, log a warning asking the user to raise ulimit for better perf.

**Comment by jbenet** [comment]
cc @diasdavid @whyrusleeping 

@spikebike i think this is why it failed for you.

**Comment by spikebike** [comment]
On my ubuntu-14.04 desktop running ipfs since June 18th:
bill@kona:~$ ps acux | grep ipfs
bill      9929  2.3  6.0 6811060 989232 pts/26 Sl+  Jun18 682:43 ipfs
bill@kona:~$ ls -al /proc/9929/fd | wc -l
77
bill@kona:~$ ulimit -n
1024

Wouldn't open files show up under /proc/<pid>/fd?

**Comment by whyrusleeping** [comment]
@spikebike yeah, this isnt the problem you were having. But it may be related, the reason that the open file issue caused the transfers to hang is because of a bug in bitswap, if your datastore was throwing an error (other than too many open files) then the same external behavior would manifest.

tldr, probably a datastore error thats not being logged, try doing: `ipfs log level bitswap debug` and reproducing the issue.

**Comment by whyrusleeping** [comment]
we have improved this both in flatfs and libp2p
